### DISCUSSION TITLE
CUDA Running out of memory Win 11 GTX 3080

### DISCUSSION THREAD

**Discussion by cmosboss** [openai/whisper#895] [Q&A] [answered]
I have been attempting to get Whisper to work with Cuda for several days now with no success. Using the CPU device works in all model sizes without issue, but I have a very powerful GPU and I want to take advantage of that. 
OS: ```Windows 11```
Python Version: ```3.7.9```
CUDA Version: ```11.6```
Language Model: ```large```
NVCC: ```/c/Program Files/NVIDIA GPU Computing Toolkit/CUDA/v11.6/bin/nvcc```

When running my code I print out ```torch.cuda.__version__``` and ```torch.cuda.is_available()``` which returns ```1.13.1+cu116``` and ```True```. 

However when I actually go to run my application I get the following error.
```
File "C:\Users\Table\.pyenv\pyenv-win\versions\3.7.9\lib\site-packages\torch\nn\modules\module.py", line 664, in _apply
    param_applied = fn(param)
  File "C:\Users\Table\.pyenv\pyenv-win\versions\3.7.9\lib\site-packages\torch\nn\modules\module.py", line 987, in convert
    return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)
RuntimeError: CUDA error: out of memory
```

This error seems really open ended and not helpful, however if I add  ``` torch.cuda.set_per_process_memory_fraction(0.5)```  (a potential fix I found online) to my script I get a slightly more detailed error message which also seems incorrect.
```
File "C:\Users\Table\.pyenv\pyenv-win\versions\3.7.9\lib\site-packages\torch\nn\modules\module.py", line 987, in convert
    return t.to(device, dtype if t.is_floating_point() or t.is_complex() else None, non_blocking)
torch.cuda.OutOfMemoryError: CUDA out of memory. Tried to allocate 20.00 MiB (GPU 0; 10.00 GiB total capacity; 4.71 GiB already allocated; 3.47 GiB free; 5.00 GiB allowed; 5.00 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```
The error message indicates I have several GB available but it cannot allocate event 20mb.

Additional info, output from nvidia-smi showing I have no phantom processes and over 7GiB of free vRam.
```
Wed Jan 25 13:00:26 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 528.02       Driver Version: 528.02       CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name            TCC/WDDM | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ... WDDM  | 00000000:08:00.0  On |                  N/A |
| 62%   57C    P0   111W / 340W |   3277MiB / 10240MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+

+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|    0   N/A  N/A      1760    C+G   ...perience\NVIDIA Share.exe    N/A      |
|    0   N/A  N/A      2604    C+G   ...in7x64\steamwebhelper.exe    N/A      |
|    0   N/A  N/A      3324    C+G   ...wekyb3d8bbwe\Video.UI.exe    N/A      |
|    0   N/A  N/A      4260    C+G   ...Battle.net\Battle.net.exe    N/A      |
|    0   N/A  N/A      5600    C+G   ...lPanel\SystemSettings.exe    N/A      |
|    0   N/A  N/A     10616    C+G   C:\Windows\explorer.exe         N/A      |
|    0   N/A  N/A     12844    C+G   ...icrosoft VS Code\Code.exe    N/A      |
|    0   N/A  N/A     13052    C+G   ...n1h2txyewy\SearchHost.exe    N/A      |
|    0   N/A  N/A     13184    C+G   ...artMenuExperienceHost.exe    N/A      |
|    0   N/A  N/A     14440    C+G   ...cw5n1h2txyewy\LockApp.exe    N/A      |
|    0   N/A  N/A     15448    C+G   ...\app-1.0.9010\Discord.exe    N/A      |
|    0   N/A  N/A     16240    C+G   ...2txyewy\TextInputHost.exe    N/A      |
|    0   N/A  N/A     16304    C+G   ...perience\NVIDIA Share.exe    N/A      |
|    0   N/A  N/A     17496    C+G   ...me\Application\chrome.exe    N/A      |
|    0   N/A  N/A     20388    C+G   ...y\ShellExperienceHost.exe    N/A      |
|    0   N/A  N/A     20424    C+G   ...e\PhoneExperienceHost.exe    N/A      |
|    0   N/A  N/A     20856    C+G   ...\Programs\Hyper\Hyper.exe    N/A      |
|    0   N/A  N/A     21692    C+G   ...518.61\msedgewebview2.exe    N/A      |
|    0   N/A  N/A     46712    C+G   ...\Programs\Hyper\Hyper.exe    N/A      |
+-----------------------------------------------------------------------------+
```

**Comment by glangford** [comment, marked as answer]
> Language Model: large
...
> Additional info, output from nvidia-smi showing I have no phantom processes and over 7GiB of free vRam.

According to the README, the large model needs ~10GB of VRAM.
https://github.com/openai/whisper#available-models-and-languages

Have you tried the small model first to see if that works?

**Reply by cmosboss** [reply]
Interesting, I had not assumed it was a total VRAM issue as the actual error indicates such a small amount _"Tried to allocate 20.00 MiB"_ Switching to a medium model does in fact work. I would assume that not having ideal amount of VRAM would just make it slower not inoperable. I also do have 10GB of VRAM available on this card but apparently there must be some overhead i.e. you need 10GB of FREE VRAM. 

Is there no way to run the whisper transcribe in batches or slower with less than 10GB VRAM? The medium model does not accurately transcribe like the large does for my use case.

**Reply by glangford** [reply]
It's worth experimenting with different transcription options; for my primary use case I found that the medium model with `--patience 2` works well, but I prefer large also.

Changes from the current version of whisper would be required to use substantially less memory, see this past discussion for example
* https://github.com/openai/whisper/discussions/180
